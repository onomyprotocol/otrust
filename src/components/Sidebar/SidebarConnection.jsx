import React from 'react';
import styled from 'styled-components';

import { ConnectionStatus } from '../UI';
import { responsive } from 'theme/constants';

const Connection = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;

  padding: 40px;

  @media screen and (max-width: ${responsive.laptop}) {
    padding: 24px;
  }

  @media screen and (max-width: ${responsive.tablet}) {
    flex-direction: row;
    gap: 40px;
  }

  @media screen and (max-width: ${responsive.smartphoneLarge}) {
    display: none;
  }
`;

const ConnectionRow = styled.div`
  display: flex;
  justify-content: space-between;

  @media screen and (max-width: ${responsive.tablet}) {
    flex-direction: column;
    justify-content: flex-start;
    gap: 10px;
  }

  > strong {
    font-weight: 400;
    color: ${props => props.theme.colors.textSecondary};
  }

  > span {
    font-weight: 500;
  }
`;

export default function SidebarConnection({ active, error, chainId, blockNumber }) {
  return (
    <Connection>
      <ConnectionRow>
        <strong>Connection</strong>
        <ConnectionStatus>{active ? 'Connected' : error ? 'Error' : 'Loading'}</ConnectionStatus>
      </ConnectionRow>
      <ConnectionRow>
        <strong>Chain Id</strong>
        <span>{chainId ?? ''}</span>
      </ConnectionRow>
      <ConnectionRow>
        <strong>Block Number</strong>
        <span>{blockNumber ? blockNumber.toFixed(0) : ''}</span>
      </ConnectionRow>
    </Connection>
  );
}
