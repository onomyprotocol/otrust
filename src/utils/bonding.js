const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
