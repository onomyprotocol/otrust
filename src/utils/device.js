const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
