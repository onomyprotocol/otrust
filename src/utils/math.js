import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
