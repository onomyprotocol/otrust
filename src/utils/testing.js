import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'Theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const renderWithContext = (Component, props) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider value={{ supplyNOM: new BigNumber(0), theme: darkNew }}>
          <ExchangeContext.Provider value={{ askAmount: 0, bidAmount: 0, bidDenom: 'strong' }}>
            <Component {...props} />
          </ExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
