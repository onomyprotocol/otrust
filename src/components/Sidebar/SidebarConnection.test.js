import { cleanup } from '@testing-library/react';

import SidebarConnection from './SidebarConnection';
import { renderWithTheme } from '../../utils/testing';

const testProps = {
  active: true,
  error: false,
  chainId: '123',
  blockNumber: 123456,
};

describe('Given the SidebarConnection component', () => {
  describe('when the component is rendered and active is true', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(SidebarConnection, testProps);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the component is rendered and error is true', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(SidebarConnection, { ...testProps, error: true });
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the component is rendered has inactive state and has no errors', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(SidebarConnection, {
        ...testProps,
        active: false,
        blockNumber: undefined,
        chainId: null,
      });
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
