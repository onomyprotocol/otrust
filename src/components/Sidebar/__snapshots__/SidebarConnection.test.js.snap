// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the SidebarConnection component when the component is rendered and active is true should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iCoGMd dWpZnl"
  >
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Connection
      </strong>
      <div
        class="sc-gKAaRy hGDlOd"
      >
        Connected
      </div>
    </div>
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Chain Id
      </strong>
      <span>
        123
      </span>
    </div>
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Block Number
      </strong>
      <span>
        123456
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Given the SidebarConnection component when the component is rendered and error is true should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iCoGMd dWpZnl"
  >
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Connection
      </strong>
      <div
        class="sc-gKAaRy hGDlOd"
      >
        Connected
      </div>
    </div>
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Chain Id
      </strong>
      <span>
        123
      </span>
    </div>
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Block Number
      </strong>
      <span>
        123456
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Given the SidebarConnection component when the component is rendered has inactive state and has no errors should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iCoGMd dWpZnl"
  >
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Connection
      </strong>
      <div
        class="sc-gKAaRy hGDlOd"
      >
        Loading
      </div>
    </div>
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Chain Id
      </strong>
      <span />
    </div>
    <div
      class="sc-fujyAs chjQuE"
    >
      <strong>
        Block Number
      </strong>
      <span />
    </div>
  </div>
</DocumentFragment>
`;
